db/db_impl/db_impl.cc

```diff
@@ -1855,6 +1856,74 @@ Status DBImpl::GetFullHistoryTsLow(ColumnFamilyHandle* column_family,
   return Status::OK();
 }
 
+Status DBImpl::GetNewestUserDefinedTimestamp(ColumnFamilyHandle* column_family,
+                                             std::string* newest_timestamp) {
+  if (newest_timestamp == nullptr) {
+    return Status::InvalidArgument("newest_timestamp is nullptr");
+  }
+  ColumnFamilyData* cfd = nullptr;
+  if (column_family == nullptr) {
+    cfd = default_cf_handle_->cfd();
+  } else {
+    auto cfh = static_cast_with_check<ColumnFamilyHandleImpl>(column_family);
+    assert(cfh != nullptr);
+    cfd = cfh->cfd();
+  }
+  assert(cfd != nullptr && cfd->user_comparator() != nullptr);
+  if (cfd->user_comparator()->timestamp_size() == 0) {
+    return Status::InvalidArgument(
+        "Timestamp is not enabled in this column family");
+  }
+
+  Status status;
+  // Acquire SuperVersion
+  SuperVersion* sv = GetAndRefSuperVersion(cfd);
+  {
+    InstrumentedMutexLock l(&mutex_);
+    bool enter_write_thread = sv->mem == cfd->mem();
+    WriteThread::Writer w;
+    // Enter write thread to read the mutable memtable to avoid racing access
```

Comment:
Would it be better to support concurrent access to newest_udt_ from memtable instead of blocking writes for each call to GetNewestUserDefinedTimestamp()?

Reply:
Yes, that's a more performant direction. Right now, the `newest_udt_` is a Slice so concurrent access is not available. For uint64_t timestamp, we can consider using std::atomic<uint64_t>.

Since the internal user for this API currently only need to call this API after DB::Open before any writes start, to get the timestamp to recover from. I think entering write thread to avoid the concurrent issue isn't a bad idea. If we need this API to be more performant, I can revise the implementation. We introduce this API for them because the `GetFullHistoryTsLow` API doesn't feel generic enough.